[Bug #1299029] phaser effect

jercaianu opened the pull request:
I tried to implement a phaser inspired by the one already done in Audacity. However the output does not sound well. I'm not sure whether this can be improved, or if i should start looking somewhere else. Thanks !

jercaianu committed:
[Bug #1299029] phaser effect

jercaianu committed:
Redone implementation with better sound output

jercaianu committed:
Changed depth limits to 0 and 1

jercaianu committed:
improved sound quality

jercaianu committed:
Rewrote phaser with much better sound output

jercaianu commented:
I found a phaser here [0] which served as a starting point. I also used the formula from [1] which is also used in the phaser listed above.
All in all, it sounds really well except some minor issues. Can someone test it?
[0] - https://code.soundsoftware.ac.uk/projects/audio_effects_textbook_code/repository/entry/effects/phaser/
[1] - https://ccrma.stanford.edu/~jos/pasp/Classic_Virtual_Analog_Phase.html

ywwg commented:
Thanks for working on this! In what way did you use the textbook's code?  Like, as a base of c++ which you then edited to match Mixxx's API, or more just for how the phaser algorithm works?

jercaianu commented:
To be honest, I found the code useful in both ways. The implementation from that book is almost identical to the one here [0] (also the ones in the comment). What is different from the implementations from there is the "magical" formula which I put above. The recipe is the same: combining the original signal with a copy passed through several all-pass filters which only change the phase. 

[0] - http://musicdsp.org/showArchiveComment.php?ArchiveID=78

jercaianu commented:
Should I try another implementation?

daschuer commented:
Thank you for your great work. 

Just tested the phaser and it sounds crazy. ... 
I think we need to tweak the scales and ranges a bit, it is too effective for me. The overall sound quality seams to be good. 

You may look at stomp box phaser reviews on youtube to find nice knob scales/ranges.

I do not like the the stereo phasing. As far as I know a classic phaser does not do this.
I would prefer to keep this off the effect and do it in a dedicated AutoPaning effect, like proposed in an other effect pull request. At least we need a control to switch it off.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,170 @@
+
+        if ((counter++) % updateCoef == 0) {
+            for (int j = 0; j < stages; j++) {
+                CSAMPLE delayLeft = 0.5 + 0.5 * sin(leftPhase);
```
This part of the code is confusing. What is delayLeft, really a sample value, or a time? 
It changes the content though the function. It would be nice to use a new speaking name every time the content changes. (The compiler will throw them out anyway)

daschuer replied:
What does sweepWidth?

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,170 @@
+                delayLeft = tan(delayLeft / 2);
+                delayRight = tan(delayRight / 2);
+
+                filterLeft[j] = (1.0 - delayLeft) / (1.0 + delayLeft);
```
This is a filter coefficient, right? Please call it like that.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,170 @@
+            }
+        }
+
+        for (int j = 0; j < stages; j++) {
```
This is the iir filter loop.
I would prefer to reformat it similar to this: 
https://github.com/mixxxdj/mixxx/blob/master/src/engine/enginefilteriir.h#L268 
Or this 
https://github.com/mixxxdj/mixxx/blob/master/src/engine/enginefiltermoogladder4.h#L151
To have a reference to the used iir model. 
This will help to recognize the form DF1 / DF2

daschuer replied:
You have "Direct Form 1"

daschuer commented:
Please add a lot more comments. 

The resonance circuit suffers clamping. This can be avoided by limit it with tanh. 
it is used here, which suffers the same: https://github.com/mixxxdj/mixxx/blob/master/src/engine/enginefiltermoogladder4.h#L204

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,170 @@
+        rightPhase = fmodf(freqSkip * pState->time + M_PI, 2.0 * M_PI);
+
+        if ((counter++) % updateCoef == 0) {
+            for (int j = 0; j < stages; j++) {
```
Why is this a loop? Didn't we use the same coefs for all stages?

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,170 @@
+
+        for (int j = 0; j < stages; j++) {
+            oldOutLeft[j] = (filterLeft[j] * left) + 
+                (filterLeft[j] * oldOutLeft[j]) - oldInLeft[j];
```
If filterLeft[j] is 0, the result is -left shifted by one sample after a second stage it is again +left shifted by two samples. Not sure if the 180 ° phase this is intended. Maybe this is the cause for the gain changes when shifting though stages.

ywwg commented:
Hey @daschuer , I'm working with @jercaianu  to find another implementation that will be free of the licensing issues here.  You can hold off on the review until we know what we're doing.

jercaianu committed:
Added stereo option and removed redundant coefficient computing

daschuer commented:
What is exactly the license issue here? I cannot find a single line copied from the GPL 3 textbook source. The underlying knowledge is in the public domain and pretty much the same was discussed at the musicdsp mailing list, which is "for anyone who is interested". But again, @jercaianu has not just copied an existing implementation from there.

ywwg commented:
That is what I was discussing with him specifically.  I didn't look through the source myself.  He seemed to agree that there could be an issue.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    EffectManifestParameter* stereo = manifest.addParameter();
+    stereo->setId("stereo");
+    stereo->setName(QObject::tr("Stereo"));
```
Thank you. This works for me for now. Maybe we can set the shift between L and R later by a knob.

jercaianu committed:
Phaser now has only even number of stages

jercaianu committed:
Tweaked some default values and added better names for variables

jercaianu committed:
Reformated sample processing part

jercaianu commented:
Thank you for your reviews! I have reformatted the code and i think the sound ranges are ok.
I'm trying to move the "Stereo" toggle button to the side of the effect rack but I don't know how to do that.
 Also, the "Stages" knob should be replaced with a different kind of knob because it designates a "discrete" variable. Any ideas?

jercaianu committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into phaser

badescunicu commented:
@jercaianu: The position of the effect parameters are decided by the skin. You can't put an effect parameter to the side of the effect rack. As far as I know, this is the order in the skin: 
```param, button_param, param, button_param, etc.``` Your parameter will go in the first available spot.

Regarding the ```Stages``` parameter: This is also a skinning issue. Currently, the buttons have only two states, so if you use ```CONTROL_TOGGLE_STEPPING``` ControlHint, it will use a two state button. You can leave it either as it is right now, as a knob, or make it use a ```CONTROL_TOGGLE_STEPPING``` ControlHint, with the maximum value set at the maximum number of available stages for the phaser, with the restriction that with the available skins and configuration, only 2 stages are "usable".

badescunicu commented:
Other than the above comments, are there any other licensing issues? What else is needed to merge this PR? (  @daschuer , @ywwg )

daschuer commented:
IMHO we have no license issues here (see my comment above)  Since we can consider this as @jercaianu's work, we can merge it at any time.

daschuer commented:
We need really MUCH more comments in this code to reduce the magical nature a bit. 
There is nothing against a text block that describes what a phaser is and how the implementation works. 
Comments above the critical lines will work for me as well.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,173 @@
+    stages->setControlHint(EffectManifestParameter::CONTROL_KNOB_LINEAR);
+    stages->setSemanticHint(EffectManifestParameter::SEMANTIC_UNKNOWN);
+    stages->setUnitsHint(EffectManifestParameter::UNITS_UNKNOWN);
+    stages->setDefault(1.0);
```
The old version was correct. I think it is common to set the stages in double steps.
2 stages = 1 notch  
You find 2/6/12 stages phaser in the wild.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,173 @@
+    CSAMPLE depth = m_pDepthParameter->value();
+    CSAMPLE feedback = m_pFeedbackParameter->value();
+    CSAMPLE range = m_pRangeParameter->value();
+    int stages = 2 * m_pStagesParameter->value();
```
You can use here a switch control like Nicu has proposed to get around the scaling issue. 
Otherwise you need a region that snaps to a desired number of stages.

jercaianu committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into phaser

daschuer commented:
There is a crackling noise when turning the Stages knob. 
I think you need to fade in and out the additional stages. Or have two sets of filters and fade from the old set to the new one.

daschuer commented:
The resonance still suffers overshooting, we should avoid to break the users speakers ;-)
Did you consider a tanh clipper?

daschuer commented:
The rate slider suffers bubbling noise when turned. Be sure the coefs do not jump.

daschuer commented:
The range slider produces sample overshoot when at leftmost position.

jercaianu committed:
Fixed bubbling noise from the 'Rate' knob, added tanh clipper and comments

jercaianu committed:
added tanh

jercaianu commented:
@daschuer  Thank you for your feedback! I fixed the rate knob issue by using the old rate value until I have settled for a new one. Also, I added tanh to where I'm using the feedback value. I have used the knob Nicu proposed. Hopefully I added enough comments.

daschuer commented:
The signal is clipped by tanh even though the depth parameter is 0. 
IMHO the effect should not touch the samples in this case. 
tanh in the feedback path should be sufficient (Works quite good in my tests)

jercaianu committed:
clipped only feedback

jercaianu commented:
Yes, i included the input signal by mistake. I fixed the issue

daschuer commented:
Thank you, the "noise" issues are gone now. :-) 

Negative range of the feedback ... OK, we have an effect in that range, but is this a phaser? 
Do you have phaser for reference that features an inverse feedback? 
This negative part of the feedback knob was confusing, at least for me. 
So should make the purpose more clear or just remove it.

LFO Rate: We should allow much faster rates like 50 or something, this would be no problem if we switch to a logarithmic knob.  

The stages knob is somehow broken. (not your fault) It switches to zero if you click on it.
Can you have a look? If not please file a bug at launchpad. 
Do you have a good idea to change the knob appearance when stepping though the values.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,199 @@
+    EffectManifestParameter* frequency = manifest.addParameter();
+    frequency->setId("lfo_frequency");
+    frequency->setName(QObject::tr("Rate"));
+    frequency->setDescription("Controls the speed of the effect.");   
```
"Controls the speed of the low frequency oscillator"

jercaianu commented:
Regarding the negative feedback, I have used as a reference the phaser in Audacity, which also offers this option. Regarding the stages knob, I shall take a look into the problem. I feel that if the rate took higher values the output sound would be too chaotic. I also think that the max rate value I used is a bit higher than necessary.

jercaianu committed:
updated frequency description

jercaianu commented:
In order for the code to be good to merge, what else should I do?
I don't want to seem pushy, but soon, I will have to present at my university a feature that I've been working on for Mixxx, and I would like to present the phaser. Thank you :)

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    if (pState->change) {
+        pState->wait++;
+        if (pState->wait == pState->limit) {
```
I do not understand how it works. Playing with the rate knob it looks like it works sometimes. 
If you move the knob continuously the new value seams to be not adopted. 
It looks like the root issue is below. See below ...

jercaianu replied:
Yes, I intended this behaviour in order to avoid bubbling noises while changing the rate knob. The rate value updates when i settle for a certain value.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    const int kChannels = 2;
+    for (unsigned int i = 0; i < numSamples; i += kChannels) {
+        pState->time++;
```
Time is growing infinite and will finally overflow. You should store the phase instead and add a phase fraction per sample. This will also avoid the bubbling noise.

daschuer commented on src/effects/native/phasereffect.cpp:
```diff
@@ -0,0 +1,199 @@
+        left = processSample(left, oldInLeft, oldOutLeft, filterCoefLeft, stages);
+        right = processSample(right, oldInRight, oldOutRight, filterCoefRight, stages);
+
+        //Computing output combining the original and processed sample
```
Nit: We use a space after // to distinguish comments from deactivated code. Please add a space here and in similar places.

daschuer commented:
There should be nothing against merging this after the pending rate issue is fixed. 
The multi state issue can be fixed later. Is it already tracked by Launchpad? If not, please file a bug for it. Do you have discovered some more about it?

jercaianu commented:
I have looked for the issue in the widget folder. I think that is issue is somewhere located in weffectpushbutton or wpushbutton, but i havent exactly pinpointed the problem.

jercaianu committed:
Fixed comment indentation and rate knob issue

jercaianu commented:
Really nice fix for the rate knob issue. I guess that it would be alright for the moment to use the previous knob for 'Stages' until the toggle one is fixed.

jercaianu committed:
Added linear knob until toggle one is fixed

jercaianu commented:
Filed a bug report for the knob issue
 https://bugs.launchpad.net/mixxx/+bug/1456357

daschuer commented:
LGTM! It works really nice now 
(except the blind fight using the stages knob) 
Thank you for all the work.
